src/dialog.mjs

```diff
@@ -39,30 +42,86 @@ const Dialog = ({ config, preferences }) => {
    */
   const dialog = htmlToElement(renderDialog());
   const form = dialog.querySelector('form');
+  const button = form.querySelector('button');
 
   /**
    * Hide/show helpers.
    */
   const hide = () => dialog.setAttribute('aria-hidden', 'true');
   const show = () => dialog.setAttribute('aria-hidden', 'false');
 
+  /**
+   * Compose values based on input type, the `acceptAllButton` configuration,
+   * any stored preferences and the state of the input options.
+   *
+   * When an option is checked, the form will just submit. When nothing is checked,
+   * it depends on the type and the `acceptAllButton` config setting.
+   *
+   * - Radio + option checked: success
+   * - Radio + nothing checked + has `acceptAllButton`: select first + success
+   * - Radio + nothing checked + no `acceptAllButton`: fail
+   *
+   * - Checkbox + option(s) checked: success
+   * - Checkbox + nothing checked + has `acceptAllButton`: select all + success
+   * - Checkbox + nothing checked + no `acceptAllButton`: success
+   */
+  const composeValues = values => {
+
+    // Radio when no option is selected.
+    if (TYPE === 'radio' && !values.find(v => v.accepted)) {
+      // If the `acceptAllButton` is option configured, select the first option and
+      // let the form submit as if the user had selected it.
+      if (ACCEPT_ALL_BUTTON) {
+        values[0].accepted = true;
+        return values;
+      }
+      // Do not submit if no option is selected and `acceptAllButton` is not configured.
+      return [];
+    }
+
+    // Checkbox with `acceptAllButton` and no user-choosable option is checked.
+    // We compare amount of required options against checked options.
+    const requiredCount = config.get('cookies').filter(c => c.required).length;
+    const checkedCount = values.filter(v => v.accepted).length;
+    const optionsChecked = requiredCount !== checkedCount;
```

Comment:
I don't fully understand the naming here. `optionsChecked` sounds like all options are checked if true, but the boolean represents the opposite.

Also, shouldn't the `if` statement below use `checkedCount` to determine whether to "check all"?
If the user has checked a single box it indicates they want to make their own choice, right?

I'm probably missing something.

Reply:
> I don't fully understand the naming here. `optionsChecked` sounds like all options are checked if true, but the boolean represents the opposite.

I had the reverse at first, but it will be a negation somewhere anyway. Either in the variable or in the if statement.

But it means that 'user options are checked', since the amount of option is different from the required amount of options. You can't be lower than the required, and if it's not the same, than it's more.

Would this make more sense?

```js
const requiredCount = config.get('cookies').filter(c => c.required).length;
const checkedCount = values.filter(v => v.accepted).length;
const userOptionsChecked = checkedCount > requiredCount;
if (ACCEPT_ALL_BUTTON && TYPE === 'checkbox' && !userOptionsChecked) {
return values.map(value => ({
  ...value,
  accepted: true,
}));
}
```

Or this? (I think I had it kind of like this at first)

```js
const requiredCount = config.get('cookies').filter(c => c.required).length;
const checkedCount = values.filter(v => v.accepted).length;
const noUserOptionsChecked = checkedCount <= requiredCount;
if (ACCEPT_ALL_BUTTON && TYPE === 'checkbox' && noUserOptionsChecked) {
  return values.map(value => ({
    ...value,
    accepted: true,
  }));
}
```

> Also, shouldn't the `if` statement below use `checkedCount` to determine whether to "check all"?

No, I don't think so? That's what the variable above is for. If it's untouched, check them all. If it's touched, then skip the `if` statement and continue returning the actual values.

> If the user has checked a single box it indicates they want to make their own choice, right?

Yes! The button changes to the default label, and it will return the user choice instead of choosing the 'highest' radio button or selecting all checkboxes.

> I'm probably missing something.

I hope so!

Reply:
Yes! I like your current version, thanks!